File: addons/GuideCS/Modifiers/GUIDEModifierScale.cs
using System;
using Godot;

namespace Guide;

[Tool]
[GlobalClass]
public partial class GUIDEModifierScale : GUIDEModifier {
	private Vector3 _scale = Vector3.One;
	[Export]
	public Vector3 Scale {
		get => _scale;
		set {
			_scale = value;
			EmitChanged();
		}
	}

	private bool _applyDeltaTime = false;
	[Export]
	public bool ApplyDeltaTime {
		get => _applyDeltaTime;
		set {
			_applyDeltaTime = value;
			EmitChanged();
		}
	}

	public override bool IsSameAs(GUIDEModifier other) {
		return other is GUIDEModifierScale otherScale &&
			   ApplyDeltaTime == otherScale.ApplyDeltaTime &&
			   Scale.IsEqualApprox(otherScale.Scale);
	}

	public override Vector3 _ModifyInput(Vector3 input, float delta, GUIDEAction.GUIDEActionValueType valueType) {
		if (!input.IsFinite()) return Vector3.Inf;

		if (ApplyDeltaTime) {
			return input * Scale * delta;
		}
		else {
			return input * Scale;
		}
	}

	public override string _EditorName() => "Scale";

	public override string _EditorDescription() => "Scales the input by the given value and optionally, delta time.";
}

File: addons/guide-cpp/src/modifiers/guide_modifier_scale.cpp
#include "guide_modifier_scale.h"

using namespace godot;

void GUIDEModifierScale::_bind_methods() {
    ClassDB::bind_method(D_METHOD("get_scale"), &GUIDEModifierScale::get_scale);
    ClassDB::bind_method(D_METHOD("set_scale", "value"), &GUIDEModifierScale::set_scale);
    ADD_PROPERTY(PropertyInfo(Variant::VECTOR3, "scale"), "set_scale", "get_scale");

    ClassDB::bind_method(D_METHOD("get_apply_delta_time"), &GUIDEModifierScale::get_apply_delta_time);
    ClassDB::bind_method(D_METHOD("set_apply_delta_time", "value"), &GUIDEModifierScale::set_apply_delta_time);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "apply_delta_time"), "set_apply_delta_time", "get_apply_delta_time");
}

GUIDEModifierScale::GUIDEModifierScale() {
}

GUIDEModifierScale::~GUIDEModifierScale() {
}

Vector3 GUIDEModifierScale::_modify_input(Vector3 input, double delta, GUIDEAction::GUIDEActionValueType value_type) const {
    if (!input.is_finite()) return Vector3(Math_INF, Math_INF, Math_INF);
    if (apply_delta_time) {
        return input * scale * delta;
    } else {
        return input * scale;
    }
}

bool GUIDEModifierScale::is_same_as(const Ref<GUIDEModifier> &other) const {
    Ref<GUIDEModifierScale> o = other;
    if (o.is_null()) return false;
    return o->get_scale().is_equal_approx(scale) &&
           o->get_apply_delta_time() == apply_delta_time;
}

String GUIDEModifierScale::_editor_name() const {
    return "Scale";
}

String GUIDEModifierScale::_editor_description() const {
    return "Scales the input by the given value and optionally, delta time.";
}